src/interface/data_collection.cpp

```diff
@@ -0,0 +1,66 @@
+//========================================================================================
+// (C) (or copyright) 2020. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#include <string>
+
+#include "interface/data_collection.hpp"
+#include "interface/mesh_data.hpp"
+#include "interface/meshblock_data.hpp"
+#include "mesh/mesh.hpp"
+#include "utils/partition_stl_containers.hpp"
+
+namespace parthenon {
+
+template <typename T>
+std::shared_ptr<T> DataCollection<T>::Add(const std::string &name,
+                                          const std::shared_ptr<T> &src) {
+  // error check for duplicate names
+  auto it = containers_.find(name);
+  if (it != containers_.end()) {
+    // check to make sure they are the same
+    if (!(*src == *(it->second))) {
+      throw std::runtime_error("Error attempting to add a Container to a Collection");
+    }
+    return it->second;
+  }
+
+  auto c = std::make_shared<T>();
+  c->Copy(src);
+
+  containers_[name] = c;
+  return containers_[name];
+}
+
+template <>
+std::shared_ptr<MeshData<Real>> &
+DataCollection<MeshData<Real>>::GetOrAdd(const std::string &mbd_label,
+                                         const int &partition_id) {
+  const std::string label = mbd_label + "_part-" + std::to_string(partition_id);
+  auto it = containers_.find(label);
+  if (it == containers_.end()) {
+    // TODO(someone) add caching of partitions to Mesh at some point
```

Comment:
Do we know if this is a performance bottleneck? Partitioning should be very fast, as it just moves some shared pointers around. Also if packs and `MeshData<Real>>` are already cached, then will the partition caching ever be relevant?

Reply:
As far as I've looked at the profiling data (without looking at this part specifically) this is now likely less of an issue.
I only added the comment as in the first iteration in this PR without the `GetOrAdd` function, the partitioning would happen in every (sub)cycle just to extract the right (cached) container.
With the current automated caching in place this should be negligible compared to other piece that are triggered when the caches are invalid (such as after load balancing).